src/com/amazon/ion/BufferConfiguration.java

```diff
@@ -0,0 +1,186 @@
+ * @param <U> the type of the concrete subclass of this BufferConfiguration that is applicable to the
+ *            ReaderLookaheadBufferBase subclass.
+ */
+public abstract class BufferConfiguration<T extends BufferEventHandler, U extends BufferConfiguration<T, U>> {
+
+    /**
+     * Provides logic common to all BufferConfiguration Builder implementations.
+     * @param <T> the type of {@link BufferEventHandler} used by the BufferConfiguration.
+     * @param <U> the type of BufferConfiguration.
+     * @param <V> the type of Builder that builds BufferConfiguration subclasses of type T.
+     */
+    public static abstract class Builder<
+        T extends BufferEventHandler,
+        U extends BufferConfiguration<T, U>,
+        V extends Builder<T, U, V>
+    > {
+
+        /**
+         * Large enough that most streams will never need to grow the buffer. NOTE: this only needs to be large
+         * enough to exceed the length of the longest top-level value plus any system values that precede it.
+         */
+        static final int DEFAULT_INITIAL_BUFFER_SIZE = 32 * 1024; // bytes
+
+        /**
+         * The initial size of the lookahead buffer, in bytes.
+         */
+        private int initialBufferSize = DEFAULT_INITIAL_BUFFER_SIZE;
+
+        /**
+         * The maximum number of bytes that will be buffered.
+         */
+        private int maximumBufferSize = Integer.MAX_VALUE;
+
+        /**
+         * The handler that will be notified when the maximum buffer size is exceeded.
+         */
+        private T eventHandler = null;
+
+        /**
+         * Sets the initial size of the buffer that will be used to hold the data between top-level values. Default:
+         * 64KB.
+         *
+         * @param initialBufferSizeInBytes the value.
+         * @return this Builder.
+         */
+        public final V withInitialBufferSize(final int initialBufferSizeInBytes) {
+            initialBufferSize = initialBufferSizeInBytes;
+            return (V) this;
+        }
+
+        /**
+         * @return the initial size of the lookahead buffer, in bytes.
+         */
+        public final int getInitialBufferSize() {
+            return initialBufferSize;
+        }
+
+        /**
+         * Sets the handler that will be notified when bytes are processed and when the maximum buffer size is exceeded
+         * (if applicable). If the maximum buffer size is finite (see {@link #withMaximumBufferSize(int)}, the handler
+         * is required to be non-null. Otherwise, the handler may be null, in which case the number of bytes processed
+         * will not be reported.
+         *
+         * @param handler the handler.
+         * @return this builder.
+         */
+        public final V withHandler(final T handler) {
+            eventHandler = handler;
+            return (V) this;
+        }
+
+        /**
+         * @return the handler that will be notified when the maximum buffer size is exceeded.
```

Comment:
Nit: the method above describes the handler in terms of two events (bytes processed, buffer size exceeded) while this one only mentions the latter. For consistency maybe go with something less specific in both places?

Reply:
Good point, generalized to "...when events occur." I think there is adequate documentation elsewhere (including in `BufferEventHandler` and `IonBufferEventHandler`) to describe which events can occur.